### PR TITLE
8307312: Replace "int which" with "int cp_index" in constantPool

### PR DESCRIPTION
**Pull request by matias9927**
Many accessors in the constant pool take an argument "int which" that is meant to represent an ambiguous index. Despite this, several methods in the API use "int which" when the argument is exclusively a constant pool index. This patch aims to rename all of these instances to "int cp_index" in order to be more clear and to distinguish methods that take constant pool indices and methods that use derived indices.

The callers have been updated to use more clear naming as well. Verified with tier 1-5 tests.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8307312](https://bugs.openjdk.org/browse/JDK-8307312): Replace "int which" with "int cp_index" in constantPool (**Enhancement** - P4)

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [eae4e38a](https://git.openjdk.org/jdk/pull/15027/files/eae4e38a8e6fbfa3ea799d84647fcbb5a155318a)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15027/head:pull/15027` \
`$ git checkout pull/15027`

Update a local copy of the PR: \
`$ git checkout pull/15027` \
`$ git pull https://git.openjdk.org/jdk.git pull/15027/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15027`

View PR using the GUI difftool: \
`$ git pr show -t 15027`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15027.diff">https://git.openjdk.org/jdk/pull/15027.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15027#issuecomment-1654557843)

### CODE REVIEW

**Review by coleenp — APPROVED** [review]
Looks good.

File: src/hotspot/share/oops/constantPool.cpp [hunk 1]
```diff
@@ -234,12 +234,12 @@ void ConstantPool::allocate_resolved_klasses(ClassLoaderData* loader_data, int n
 void ConstantPool::initialize_unresolved_klasses(ClassLoaderData* loader_data, TRAPS) {
   int len = length();
   int num_klasses = 0;
-  for (int i = 1; i <len; i++) {
-    switch (tag_at(i).value()) {
+  for (int cp_index = 1; cp_index <len; cp_index++) {
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/constantPool.cpp:237]: I don't think the use of `i` as a simple loop variable caused any problem in understanding here.

File: src/hotspot/share/oops/constantPool.cpp [hunk 2]
```diff
@@ -453,7 +453,7 @@ int ConstantPool::cp_to_object_index(int cp_index) {
   return (i < 0) ? _no_index_sentinel : i;
 }
 
-void ConstantPool::string_at_put(int which, int obj_index, oop str) {
+void ConstantPool::string_at_put(int obj_index, oop str) {
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/constantPool.cpp:456]: Should `obj_index` be `cp_index` here?

File: src/hotspot/share/oops/constantPool.cpp [hunk 3]
```diff
@@ -1612,27 +1612,27 @@ void ConstantPool::copy_operands(const constantPoolHandle& from_cp,
 // Copy this constant pool's entries at start_i to end_i (inclusive)
 // to the constant pool to_cp's entries starting at to_i. A total of
 // (end_i - start_i) + 1 entries are copied.
-void ConstantPool::copy_cp_to_impl(const constantPoolHandle& from_cp, int start_i, int end_i,
-       const constantPoolHandle& to_cp, int to_i, TRAPS) {
+void ConstantPool::copy_cp_to_impl(const constantPoolHandle& from_cp, int start_cpi, int end_cpi,
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/constantPool.cpp:1615]: The comment needs updating now the parameter names have changed.

Again I don't think there was any need to change the names in these cases.

File: src/hotspot/share/oops/constantPool.cpp [hunk 4]
```diff
@@ -942,7 +942,7 @@ BasicType ConstantPool::basic_type_for_constant_at(int which) {
 // Some constant pool entries cache their resolved oop. This is also
 // called to create oops from constants to use in arguments for invokedynamic
 oop ConstantPool::resolve_constant_at_impl(const constantPoolHandle& this_cp,
-                                           int index, int cache_index,
+                                           int cp_index, int cache_index,
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/constantPool.cpp:945]: Not sure that we really need to always say `cp_index` given the context. Changing `which` is good, but `index` was perfectly fine IMO.

**Review by dholmes-ora — APPROVED** [review]
The basic `which` change is good and useful, but I don't think most of the other renamings were really necessary. It is hard to maintain a convention that an index into to the CP must always be called `cp_index` - the `cp` is often redundant given the context.

Please double-check all comments where parameter names have changed to ensure they refer to the new names.

Thanks.

**Review by iklam — APPROVED** [review]

**Review by coleenp — COMMENTED** [review]
I hate to be disagreeable but the change to cp_index in resolve_constant_at_impl was good because cache_index is also a parameter in that function.  Knowing which index is very helpful there.  Can you change back just that one?

**Review by coleenp — APPROVED** [review]
Thanks for reverting resolve_constant_at_impl.

**Review by iklam — APPROVED** [review]
New update looks good.

**Comment by matias9927** [comment]
Thank you for the reviews @coleenp, @dholmes-ora, and @iklam!
/integrate

**Comment by dholmes-ora** [comment]
Thanks for rolling back some of those changes.

**Pull request closed without merging**